File: Algorithms/cpp/0020 Valid Parentheses/Valid_Parentheses.cpp
// Problem :    https://leetcode.com/problems/valid-parentheses/
// Solution:    https://github.com/tjzhym/LeetCode-solution
// Author :     zhym (tjzhym)
// Date   :     2021-7-29


#include <stack>
#include <string>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> parenthesesStack;
        for (char& c : s) {
            if (c == '(')
                parenthesesStack.push(')');
            else if (c == '[')
                parenthesesStack.push(']');
            else if (c == '{')
                parenthesesStack.push('}');
            else {
                if (parenthesesStack.empty() || parenthesesStack.top() != c)
                    return false;
                else
                    parenthesesStack.pop();
            }
        }
        return parenthesesStack.empty();
    }
};

File: Algorithms/LC/Python/0020 Valid Parentheses/Valid_Parentheses.py
# Problem :    https://leetcode.com/problems/valid-parentheses/
# Solution:    https://github.com/tjzhym/LeetCode-solution
# Author :     zhym (tjzhym)
# Date   :     2021-7-29


class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        PARENTHESES = {"[":"]", "{":"}", "(":")"}
        for char in s:
            if char in PARENTHESES.keys():
                stack.append(PARENTHESES[char])
            elif stack == [] or char != stack.pop():
                return False
        return stack == []